File: 047.py

def factors(num, primes):
    n = num
    s = set()
    size = len(primes)
    i = 0
    while True:
        p = primes[i]
        tmp = 1
        while n % p == 0:
            n /= p
            tmp *= p
        if tmp != 1:
            s.add(tmp)
        if n == 1:
            break
        else:
            i += 1
    return s


    
def is_distinct(sets, s, n):

    if len(s) != n:
        return False
        
    for ss in sets:
        if len(ss) != n:
            return False
        for x in s:
            if x in ss:
                return False
                
    return True

    

with open("prime_9.txt") as f:
    primes = [int(x) for x in f.readlines()]


print "read!"


sets = []
n1 = n2 = 0


sets.append(factors(2, primes))
n1 = 2

count = 0

for i in range(3, 10000000):

    s = factors(i, primes)
    
    if is_distinct(sets, s, 4):
        sets.append(s)
        
    else:
        sets = [s]

        
    if len(sets) == 4:
        print i - 3
        break
        

print "done"

File: 047.cpp
#include <cstdio>
#include <cmath>
#include <vector>
#include <set>
#include "reader.hpp"


std::set<int> factor(int num, Reader<int> &r){
    
    int tmp = num;
    std::set<int> fac;
    int size = r.size();
    int d;

    printf("factor: ");
    for (int i = 0; i < size; i++) {
        d = r.at(i);
        if (tmp % d == 0) {
            printf("%d, ", d);
            fac.insert(d);
            tmp /= d;
            if (tmp == 1) {
                break;
            }
            i--;
            continue;
        }
    }
    printf("\n");
    return fac;
}

template <class T>
bool is_distinct(std::set<T> &v1, std::set<T> &v2){
    // for (int i=0; i < v1.size(); i++) {
    //     if (v2.find(v1[i]) == v2.end()) return false;
    // }
    for (std::set<int>::iterator i = v1.begin(); i != v1.end(); i++) {
        if (v2.find(*i) != v2.end()) return false;
    }
    return true;
}


int main(){

    Reader<int> r("prime_9.txt");

    int n1 = 0;
    int n2 = 0;
    std::set<int> v1, v2;
    v1 = factor(1, r);

    int count = 0;

    for (int i=2; i < 100; i++) {
        printf("i: %d, ", i);
        v2 = factor(i, r);
        if (is_distinct<int>(v2, v2, 2)) {
            count++;
            n2 = i;
        }
        if (count == 1) {
            break;
        }

        n1 = i;
        v1 = v2;
    }


    
    return 0;
}
